modules/sdk-lib-mpc/src/tss/ecdsa/noSmallFactorsProof.ts

```diff
@@ -0,0 +1,119 @@
+import { createHash } from 'crypto';
+import { bitLength, randBetween } from 'bigint-crypto-utils';
+import { modPow } from 'bigint-mod-arith';
+import { bigIntFromBufferBE, bigIntToBufferBE } from '../../util';
+
+const ORDER = BigInt('0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141');
+const ELL = BigInt(256);
+const EPSILON = BigInt(BigInt(2) * ELL);
+
+/**
+ * Generate pseudo-random challenge value $e$ for $(N, w)$.
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint} - challenge value $e$.
+ */
+function generateE(N, w) {
+  const digest = createHash('shake256', { outputLength: 1 + Math.floor((bitLength(ORDER) + 7) / 8) })
+    .update(bigIntToBufferBE(N))
+    .update('$')
+    .update(bigIntToBufferBE(w))
+    .update('$')
+    .digest();
+  const e = bigIntFromBufferBE(digest.subarray(1)) % ORDER;
+  if (digest[0] & 1) {
+    return -e;
+  }
+  return e;
+}
+
+/**
+ * Calculate the closest integer square root of $n$.
+ * @param n - the number to calculate the square root of.
+ * @returns {bigint} - $n$'s closest integer square root.
+ */
+function isqrt(n: bigint): bigint {
+  if (n < BigInt(0)) {
+    throw new Error();
+  }
+  if (n < BigInt(2)) {
+    return n;
+  }
+  function newtonIteration(n, x0) {
+    const x1 = (n / x0 + x0) >> BigInt(1);
+    if (x0 === x1 || x0 === x1 - BigInt(1)) {
+      return x0;
+    }
+    return newtonIteration(n, x1);
+  }
+  return newtonIteration(n, BigInt(1));
+}
+
+/**
+ * Prove that $n0$ has no small factors, where $n0$ is the product of two large primes.
+ * @param n0 - a modulus that is the product of $p$ and $q$.
+ * @param p - the larger prime factor of $n$.
+ * @param q - the lesser prime factor of $n$.
+ * @param w - a random number with the same bitLength as $n0$, that satisfies the Jacobi of w is -1 wrt $n0$.
+ * @param nHat - a safe bi-prime, such as that returned from rangeProof.generateNTilde.
+ * @param s - security parameters for $nHat$ such as the $h1$ value returned from rangeProof.generateNTilde.
+ * @param t - security parameters for $nHat$ such as the $h2$ value returned from rangeProof.generateNTilde.
+ * @returns proof of $n0$ having no small factors.
+ */
+export function prove(n0: bigint, p: bigint, q: bigint, w: bigint, nHat: bigint, s: bigint, t: bigint) {
```

Comment:
Please define a type for this return value 

Suggestion define `DeserializedNoSmallFactorProof` in `src/tss/ecdsa/types`. 
```typescript
type NoSmallFactorProof<T> = {
	P: T,
	Q: T,
...
};
export DeserializedNoSmallFactorProof = NoSmallFactorProof<bigint>;
export SerializedNoSmallFactorProof = NoSmallFactorProof<string>;
```

Reply:
done